### PR TITLE
Lobby Changes Round 2

### PR DESCRIPTION
**Pull request by AustinKellar**
Contributors: Austin

Goal:
- Place blocks inside lobby
- Implement an exit that allows players to despawn and return to the title menu

Note:
- Exit sign art and Change Color icon is still temp

What Changed:
- LobbyBlockSpawner:
    - A simple block spawner that spawns when there are fewer than 3 blocks (this value is configurable)
        - Can toggle off respawning with a checkbox if we so choose
- LobbyBlockDestroyer
    - A simple script that goes on a trigger collider which destroys blocks upon touch
        - Used to make sure blocks don't get on the ready stage and the exit area
- SquishDetector, BlockPusher, PlayerAdjacencyTracker, BlockDamager
    - Made small changes so that blocks can be pushed in the Lobby
- LobbyStateManager, PlayerSpawner
    - Added functionality so that the person who presses Play spawns into the lobby
- PlayerExitCollider
    - Despawns players and returns to title menu if the last player leaves the lobby

How to Test:
- Load up the lobby
- When you press play, you should spawn in
- Try pushing blocks around. They should behave as normal
- When there are 3 or fewer blocks in the scene, more should spawn in
- Try dying. You should respawn
- Try going to the exit. If there is more than 1 player, you should be despawned. If you are the last player, you should despawn and be taken back to the title screen
- Test in a build.

### CODE REVIEW

File: Assets/Scripts/Blocks/BlockPusher.cs [hunk 1]
```diff
@@ -97,11 +97,24 @@ public virtual bool Push(Vector3 direction, float animationDelay = 0f)
...
+                from obj in adjacentObjects
+                where obj.tag == "Block"
+                select obj.GetComponent<BlockPusher>()
+            ).ToList();
+        }
+        else
+        {
+            adjacentBlocks = (
+                from obj in adjacentObjects
+                where Spawner.Instance.BlockTagList.Contains(obj.tag)
```
Comment (IanGRap) [inline Assets/Scripts/Blocks/BlockPusher.cs]: maybe these scripts should have a privately assigned list of block tags that defaults to just block if the spawner is null. I don't think we have time for the refactor though.

File: Assets/Scripts/Player Selection/LobbyBlockSpawner.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    private GameObject _blockPrefab;
+
+    [SerializeField]
+    private bool _blockRespawningEnabled = true;
+
+    [SerializeField]
+    private int _respawnThreshold = 3;
+
+	private void Start ()
+    {
```
Comment (IanGRap) [inline Assets/Scripts/Player Selection/LobbyBlockSpawner.cs]: awkward curly brace

File: Assets/Scripts/Player Selection/LobbyBlockSpawner.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        StartCoroutine(SpawnBlocks());
+	}
+
+    private IEnumerator SpawnBlocks()
```
Comment (IanGRap) [inline Assets/Scripts/Player Selection/LobbyBlockSpawner.cs]: should this coroutine even be called if _blockRespawningEnabled is false?

File: Assets/Scripts/Player Selection/LobbyBlockSpawner.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                SpawnNewFormation();
+            }
+
+            yield return new WaitForSeconds(5f);
+        }
+    }
+
+    private void SpawnNewFormation()
+    {
+        GameObject tempFormationObject = Instantiate(GetRandomBlockFormation());
```
Comment (IanGRap) [inline Assets/Scripts/Player Selection/LobbyBlockSpawner.cs]: you dont need to instantiate the outline object. You can run the forloop on the prefab

File: Assets/Scripts/Player Selection/LobbyBlockSpawner.cs (hunk 2 above)
Comment (AustinKellar) [reply Assets/Scripts/Player Selection/LobbyBlockSpawner.cs]: It looks right in VS. I think its GitHub's interpreter being awkward.

File: Assets/Scripts/Player Selection/LobbyBlockSpawner.cs (hunk 3 above)
Comment (AustinKellar) [reply Assets/Scripts/Player Selection/LobbyBlockSpawner.cs]: Good point.

File: Assets/Scripts/Player Selection/LobbyBlockSpawner.cs (hunk 4 above)
Comment (AustinKellar) [reply Assets/Scripts/Player Selection/LobbyBlockSpawner.cs]: Fixed.

File: Assets/Scripts/Blocks/BlockPusher.cs (hunk 1 above)
Comment (AustinKellar) [reply Assets/Scripts/Blocks/BlockPusher.cs]: Yeah, if we have time to refactor this, I will. I think its out of scope for this PR.

**Comment by AustinKellar** [comment]
Alright, I implemented most of your suggestions. Block spawn locations are more contstrained and there are only 2 (small) block formations allowed. Since there is another lobby PR coming up, I will save all visual upgrades for that time.

**Pull request merged**